### PR TITLE
Make audit output version configurable.

### PR DESCRIPTION
**Pull request by crassirostris**
This is a re-make of https://github.com/kubernetes/kubernetes/pull/51786, taken over form @soltysh 

Copying from the previous PR:

This is followup to https://github.com/kubernetes/kubernetes/pull/51719 to start the discussion how we want to solve the problem of users picking which version is being served them.

We need to have an option for log and webhook, separately. Probably, for webhook backend with multiple destinations we'd like to send different version to each.

This approach adds two flags (only the second commit matters), one for log and another for webhook (unfortunately global one). I've looked into kubeconfig types and although there are options to specify group and version they are meant for removal. @liggitt had some thoughts maybe he could share the ideas and we can pick it up here.

@ericchiang @CaoShuFeng @sttts opinions, thoughts are more than welcome

```release-note
Add apiserver configuration option to choose audit output version.
```

### CODE REVIEW

**Comment by ericchiang** [comment]
/retest

I think this is a reasonable change. I don't think there's a good kubeconfig field to express this as (nothing like a "preferred version" field), so a flag seems like a good place to put the configuration.

**Comment by yliaog** [comment]
i guess i'm missing some context here. what is the reason to keep v1alpha1?

**Comment by crassirostris** [comment]
@yliaog there might be some feature in the future that will only be present in v1alpha1 API. If someone wishes to use these new feature, they will just use v1alpha1 API as output

**Comment by crassirostris** [comment]
Something doesn't work, I'm not quite sure what and there's no time

Since it's not a critical thing, moving it to 1.11 milestone

**Comment by CaoShuFeng** [comment]
This idea looks good to me.
:+1:

**Comment by crassirostris** [comment]
@sttts Could you please review when you have some time?

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go [hunk 1]
```diff
@@ -186,6 +196,31 @@ func validateBackendBatchOptions(pluginName string, options AuditBatchOptions) e
...
+	auditv1beta1.SchemeGroupVersion,
+}
+
+func validateGroupVersionString(groupVersion string) error {
+	gv, err := schema.ParseGroupVersion(groupVersion)
+	if err != nil {
+		return err
+	}
+	if !knownGroupVersion(gv) {
+		return fmt.Errorf("invalid group version, allowed versions are %q", knownGroupVersions)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: how does %q look like with an array?

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go [hunk 2]
```diff
@@ -287,6 +322,8 @@ func (o *AuditLogOptions) AddFlags(fs *pflag.FlagSet) {
 		"Format of saved audits. \"legacy\" indicates 1-line text format for each event."+
 			" \"json\" indicates structured json format. Requires the 'AdvancedAuditing' feature"+
 			" gate. Known formats are "+strings.Join(pluginlog.AllowedFormats, ",")+".")
+	fs.StringVar(&o.GroupVersionString, "audit-log-version", o.GroupVersionString,
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: audit-log-group-version

Comment (soltysh) [reply staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: I'd keep it short `audit-log-version`. Everyone familiar with k8s will expect version to be group + actual version

**Review by soltysh — APPROVED** [review]
The code lgtm, but I'll defer to Stefan for final word

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: https://play.golang.org/p/38dO_hlw0K5: `["foo" "bar"]`

**Pull request merged**